### PR TITLE
Add support for starter projects in DevWorkspaces

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
See https://github.com/devfile/devworkspace-operator/pull/1130#issuecomment-1604925774 for additional information; initially starterProject's `subDir` field was handled incorrectly.

Add attribute `controller.devfile.io/use-starter-project` (naming suggestions welcome) which can be applied in the top-level attributes of a devfile/DevWorkspace to select a starter project from the `.spec.template.starterProjects` array. If a starterProject is selected in this way, it is converted internally (within project-clone) to a regular project and processed alongside any other projects defined in the DevWorkspace. 

In order to support the `subDir` field on starterProjects, project-clone can now also perform sparse checkouts (though this is still not user-facing for regular projects at the moment).

I also simplified the git operations in the project clone directory -- we don't need to change dirs before running commands and can instead use the `-C` commandline option to `git`. 

A future improvement would be to expose `selectedStarterProject` as a DevWorkspace template field, rather than relying on the attribute.

### What issues does this PR fix or reference?
Closes https://github.com/devfile/devworkspace-operator/issues/1084

### Is it tested? How?
Changes to the project-clone container are pushed to `quay.io/amisevsk/project-clone:starter-projects`

PR contains a devfile useful for basic smoke testing of starter projects support: `project-clone/test/starter-project-test.devworkspace.yaml`

For manual verification, various devfiles in https://registry.devfile.io/viewer contain starter projects (e.g. [Go](https://registry.devfile.io/devfiles/go/1.0.2), which is a basic example, or [.NET 5.0](https://registry.devfile.io/devfiles/dotnet50/1.0.3) which includes a subDir. To make testing devfiles from the registry a little simpler, you can use the `yq` script:
```bash
YQ_SCRIPT='{
    "kind": "DevWorkspace",
    "apiVersion": "workspace.devfile.io/v1alpha2",
    "metadata": {
      "name": .metadata.name
    },
    "spec": {
      "started": true,
      "routingClass": "basic",
      "template": (. | del(.metadata, .schemaVersion))
    }
  }'
curl <raw-devfile> | yq -y "$YQ_SCRIPT" | kubectl apply -f -
```
(you will need to add the `controller.devfile.io/use-starter-project` attribute to actually select a project)

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

**Review by AObuchow — APPROVED** [review]
Minor comments (mainly typo's) but looks good to me and works as expected.

I tested using `subDir` with the following devworkspace, based on the dotnet 5.0 sample (with slight modifications to the container image being used to include git, and the project checkoutFrom revision):

```YAML
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: dotnet50
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/storage-type: ephemeral
      controller.devfile.io/use-starter-project: dotnet50-example
    starterProjects:
      - name: dotnet50-example
        git:
          checkoutFrom:
            remote: origin
            revision: dotnet-5.0
          remotes:
            origin: https://github.com/redhat-developer/s2i-dotnetcore-ex
        subDir: app
    components:
      - name: dotnet
        container:
          image: quay.io/wto/web-terminal-tooling:next
          args: ["tail", "-f", "/dev/null"]
          mountSources: true
          env:
            - name: CONFIGURATION
              value: Debug
            - name: STARTUP_PROJECT
              value: app.csproj
            - name: ASPNETCORE_ENVIRONMENT
              value: Development
            - name: ASPNETCORE_URLS
              value: http://*:8080
          endpoints:
            - name: http-dotnet50
              targetPort: 8080
    commands:
      - id: build
        exec:
          workingDir: ${PROJECT_SOURCE}
          commandLine: kill $(pidof dotnet); dotnet build -c $CONFIGURATION $STARTUP_PROJECT /p:UseSharedCompilation=false
          component: dotnet
          group:
            isDefault: true
            kind: build
      - id: run
        exec:
          workingDir: ${PROJECT_SOURCE}
          commandLine: dotnet run -c $CONFIGURATION --no-build --project $STARTUP_PROJECT --no-launch-profile
          component: dotnet
          group:
            isDefault: true
            kind: run
```

The workspace started up correctly. When I navigated to the project directory and did a git status, it showed that a sparse checkout was being used:

```bash
bash-4.4 /projects/dotnet50-example $ git status
On branch dotnet-5.0
Your branch is up to date with 'origin/dotnet-5.0'.

You are in a sparse checkout with 11% of tracked files present.

nothing to commit, working tree clean
```

The project clone container logs were as follows:

```
│ project-clone 2023/06/22 00:22:17 Using temporary directory /projects/project-clone-1697882419                                                                                                                                                                                                                                                      │
│ project-clone 2023/06/22 00:22:17 Read DevWorkspace at /devworkspace-metadata/flattened.devworkspace.yaml                                                                                                                                                                                                                                           │
│ project-clone 2023/06/22 00:22:17 Processing project dotnet50-example                                                                                                                                                                                                                                                                               │
│ project-clone 2023/06/22 00:22:17 Cloning project dotnet50-example to /projects/project-clone-1697882419/dotnet50-example                                                                                                                                                                                                                           │
│ project-clone 2023/06/22 00:22:17 DOING A GIT SPARSE CHECKOUT SHELL COMMAND                                                                                                                                                                                                                                                                         │
│ project-clone Cloning into '/projects/project-clone-1697882419/dotnet50-example'...                                                                                                                                                                                                                                                                 │
│ project-clone 2023/06/22 00:22:19 Cloned project dotnet50-example to /projects/project-clone-1697882419/dotnet50-example                                                                                                                                                                                                                            │
│ project-clone 2023/06/22 00:22:19 Setting up sparse checkout for project dotnet50-example                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        │
│ project-clone 2023/06/22 00:22:19 Setting up remotes for project dotnet50-example                                                                                                                                                                                                                                                                   │
│ project-clone 2023/06/22 00:22:20 Fetched remote origin at https://github.com/redhat-developer/s2i-dotnetcore-ex                                                                                                                                                                                                                                    │
│ project-clone 2023/06/22 00:22:20 Checking out remote branch dotnet-5.0                                                                                                                                                                                                                                                                             │
│ project-clone Switched to a new branch 'dotnet-5.0'                                                                                                                                                                                                                                                                                                 │
│ project-clone branch 'dotnet-5.0' set up to track 'origin/dotnet-5.0'.                                                                                                                                                                                                                                                                              │
│ project-clone 2023/06/22 00:22:20 Moving cloned project dotnet50-example from temporary dir /projects/project-clone-1697882419/dotnet50-example to /projects/dotnet50-example 
```

One note that could be for a future improvement, or as part of this PR (though solving this seems like it'd take more effort than it's worth): if you specify an invalid `subDir` for the `starterProject`, the project will still be cloned and no errors/warnings will be logged.
However, any directories belonging to the git repo won't be cloned.

For example, setting `subDir` to some non-existing directory name with the `dotnet50` devworkspace that I provided will result in `app` folder [from the repository](https://github.com/redhat-developer/s2i-dotnetcore-ex/tree/dotnet-7.0/app) being missing, as shown:

```bash
bash-4.4 /projects/dotnet50-example $ ls -la
total 32
drwxr-sr-x. 1 user user   148 Jun 22 00:23 .
drwxrwsrwx. 1 root user    32 Jun 22 00:23 ..
-rw-r--r--. 1 user user   328 Jun 22 00:23 .editorconfig
drwxr-sr-x. 1 user user   188 Jun 22 00:24 .git
-rw-r--r--. 1 user user   102 Jun 22 00:23 .gitattributes
-rw-r--r--. 1 user user    83 Jun 22 00:23 .mailmap
-rw-r--r--. 1 user user  2568 Jun 22 00:23 Contributing.adoc
-rw-r--r--. 1 user user 11357 Jun 22 00:23 LICENSE
-rw-r--r--. 1 user user  1918 Jun 22 00:23 README.adoc
bash-4.4 /projects/dotnet50-example $ git status
On branch dotnet-5.0
Your branch is up to date with 'origin/dotnet-5.0'.

You are in a sparse checkout with 11% of tracked files present.

nothing to commit, working tree clean
```

Solving this would probably require validating that `subDir` exisits as a directory within the repo, and then falling back to a normal (non-sparse) git clone if the `subDir` doesn't exist in the repo.

**Comment by AObuchow** [comment]
Sidenote: I was a bit confused about how `subDir` was supposed to work, but then found https://github.com/devfile/api/issues/128.

**Comment by amisevsk** [comment]
Re-reading the subDir documentation after @AObuchow pointed at that issue, I realized that the approach in this PR for subDir was actually incorrect.

I've pushed the following changes:
* Keep sparse checkout behavior (since it's already implemented); rename attribute to `sparseCheckout`
* Add correct behavior for subDir: copy only a subdirectory from the clone repository to project's clone path

@AObuchow please verify the updated behaviour for subDir when you have a moment. The new behavior should be
* Project is "set up", contents of project are not a git repository and contain whatever was in the subdirectory specified in `subDir`.

File: project-clone/internal/git/operations.go [hunk 1]
```diff
@@ -62,16 +63,41 @@ func CloneProject(project *dw.Project, projectPath string) error {
...
-	// Delegate to standard git binary because git.PlainClone takes a lot of memory for large repos
-	err := shell.GitCloneProject(defaultRemoteURL, defaultRemoteName, projectPath)
-	if err != nil {
-		return fmt.Errorf("failed to git clone from %s: %s", defaultRemoteURL, err)
+	if project.Attributes.Exists(internal.ProjectSubDir) {
+		if err := shell.GitSparseCloneProject(defaultRemoteURL, defaultRemoteName, projectPath); err != nil {
+			return fmt.Errorf("failed to sparsely git clone from %s: %s", defaultRemoteURL, err)
+		}
+	} else {
+		// Delegate to standard git binary because git.PlainClone takes a lot of memory for large repos
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply project-clone/internal/git/operations.go:71]: I think the comment is still useful as it warns of a major pitfall in switching to something like `PlainClone` (namely, dramatically higher memory usage).

**Review by AObuchow — APPROVED** [review]
Looks good to me, see minor comment :+1: 

Tested on OpenShift 4.12. During normal behaviour of subDir, the correct directory was put into the project folder and the directory wasn't a git repo, as expected.

I also tried giving an invalid/non-existent subdirectory, which resulted in the following logs (which were also placed into `/projects/project-clone-errors.log`:

```
2023/06/26 22:19:02 Using temporary directory /projects/project-clone-1166241122
2023/06/26 22:19:02 Read DevWorkspace at /devworkspace-metadata/flattened.devworkspace.yaml
2023/06/26 22:19:02 Processing project dotnet50-example
2023/06/26 22:19:02 Cloning project dotnet50-example to /projects/project-clone-1166241122/dotnet50-example
Cloning into '/projects/project-clone-1166241122/dotnet50-example'...
2023/06/26 22:19:02 Cloned project dotnet50-example to /projects/project-clone-1166241122/dotnet50-example
2023/06/26 22:19:02 Setting up remotes for project dotnet50-example
2023/06/26 22:19:03 Fetched remote origin at https://github.com/redhat-developer/s2i-dotnetcore-ex
2023/06/26 22:19:03 Checking out remote branch dotnet-5.0
Switched to a new branch 'dotnet-5.0'
branch 'dotnet-5.0' set up to track 'origin/dotnet-5.0'.
2023/06/26 22:19:03 Moving subdirectory invalid in project dotnet50-example from temporary directory to /projects/dotnet50-example
2023/06/26 22:19:03 Encountered error while setting up project dotnet50-example: failed to move subdirectory of cloned project to /projects: rename /projects/project-clone-1166241122/dotnet50-example/invalid /projects/dotnet50-example: no such file or directory
```

**Comment by AObuchow** [comment]
I also tested using sparseCheckout as a project attribute, which worked as expected

File: project-clone/internal/git/operations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AObuchow) [reply project-clone/internal/git/operations.go:71]: Makes sense to me :+1:

File: project-clone/internal/git/setup.go [hunk 2]
```diff
@@ -92,3 +91,27 @@ func setupRemotesForExistingProject(project *dw.Project) error {
...
 	return nil
 }
+
+func copyProjectFromTmpDir(project *dw.Project, tmpClonePath string) error {
+	if project.Attributes.Exists(internal.ProjectSubDir) {
+		// Only want one directory from the project
+		var err error
+		subDirSubPath := project.Attributes.GetString(internal.ProjectSubDir, &err)
+		if err != nil {
+			return fmt.Errorf("failed to projects subDir on project: %w", err)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply project-clone/internal/git/setup.go]: Definitely a melted brain moment :D 

Thanks!

**Pull request merged**